File: external/grpc/python/aiddl_external_grpc_python/function/function_client.py
import grpc

import aiddl_external_grpc_python.generated.sender_pb2_grpc as sender_pb2_grpc
from aiddl_external_grpc_python.converter import Converter


class FunctionClient:
    def __init__(self, host, port, container):
        channel = grpc.insecure_channel(f'{host}:{port}')
        self.stub = sender_pb2_grpc.SenderStub(channel)
        self.converter = Converter(container)

    def __call__(self, x):
        self.converter.pb2aiddl(self.stub.call(self.converter.aiddl2pb(x)))

File: external/grpc/scala/src/main/scala/org/aiddl/external/grpc/scala/function/FunctionClient.scala
package org.aiddl.external.grpc.scala.function

import io.grpc.ManagedChannelBuilder
import org.aiddl.external.grpc.function.FunctionGrpc
import org.aiddl.core.scala.container.Container
import org.aiddl.core.scala.parser.Parser
import org.aiddl.core.scala.representation.Term
import org.aiddl.core.scala.util.StopWatch
import org.aiddl.external.grpc.scala.converter.Converter

class FunctionClient(host: String, port: Int, c: Container) {
  private val converter = new Converter(c)

  private val channel = ManagedChannelBuilder
    .forAddress(host, port)
    .usePlaintext()
    .build

  private val blockingStub = FunctionGrpc.blockingStub(channel)

  def call(x: Term): Term = {
    val answer = blockingStub.call(this.converter.aiddl2pb(x))
    this.converter.pb2aiddl(answer)
  }
}